01-data-structures/03-linked-lists/linked_list.rb

```diff
@@ -6,25 +6,118 @@ class LinkedList
 
   # This method creates a new `Node` using `data`, and inserts it at the end of the list.
   def add_to_tail(node)
+    node.next = nil
+    if @tail
+      @tail.next = node
+      @tail = node
+    else
+      @tail = node
+      @head = @tail
+    end
   end
 
   # This method removes the last node in the lists and must keep the rest of the list intact.
   def remove_tail
+    if @head.next
+      temp = @head
+      temp_next = @head.next
+      while temp_next.next != nil
+        temp = temp.next
+        temp_next = temp.next
+      end
+      @tail = temp
+      temp.next = nil
+    else
+      @head = nil
+      @tail = nil
+    end
   end
 
   # This method prints out a representation of the list.
   def print
+    temp = @head
+    temp_next = @head.next
+    puts temp.data
+    puts temp_next.data
+    while temp_next.next
+      temp = temp.next
+      temp_next = temp.next
+      puts temp_next.data
+    end
   end
 
   # This method removes `node` from the list and must keep the rest of the list intact.
   def delete(node)
+    if @head == node
+      if @head.next == nil
+        @tail = nil
+      end
+      @head = @head.next
+    else
+      current_node = @head
+      next_node = @head.next
+      while next_node != node
+        current_node = next_node
+        next_node = current_node.next 
+      end
+      delete_next_node(current_node, next_node)
+    end
   end
 
   # This method adds `node` to the front of the list and must set the list's head to `node`.
   def add_to_front(node)
+    if @head
+      old_head = @head
+      @head = node
+      @head.next = old_head
+    else
+      @head = node
+      @tail = @head
+    end
   end
 
   # This method removes and returns the first node in the Linked List and must set Linked List's head to the second node.
   def remove_front
+    if @head
+      @head = @head.next
+      if !@head
+        @tail = @head
+      end
+    end
   end
-end
\ No newline at end of file
+  
+  # added this due to the benchmark portion of the assignment asking for me to delete the nth element in the List.
+  # In order to do a like for like with an Array, thought appropriate to add this method.
+  def delete_node_by_index(index)
+    if index == 0
+      if @head.next != nil
+        @head = @head.next
+      else
+        @tail = nil
+        @head = nil
+      end
+    else
+      current_node = locate_node_by_index(index - 1)
```

Comment:
How would I make this less weird/hacky? I refactored this a bit in the next commit. Let me know if this looks better.